org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/typesystem/internal/ExpressionAwareStackedResolvedTypes.java

```diff
@@ -55,5 +67,42 @@ protected void tryResolveUnboundReferences() {
 			}
 		}
 	}
+	
+	@Override
+	protected void performMergeIntoParent() {
+		if (canBeForwardResolved()) {
+			/*
+			 * XExpressions that can be forward-resolved will be kept around in the forwardComputedChildren map.
+			 * 
+			 * There are two cases: Either we performed the computation for the first time or we are
+			 * in a subsequent iteration where we used refined results from a previous run.
+			 */
+ 			IApplicableCandidate candidate = basicGetLinkingMap().get(expression);
+			if (candidate instanceof AbstractPendingLinkingCandidate<?>) {
+				AbstractPendingLinkingCandidate<?> casted = (AbstractPendingLinkingCandidate<?>) candidate;
+				if (casted.description instanceof BucketedEObjectDescription || casted.description instanceof SimpleIdentifiableElementDescription) {
+					forwardLinking().put((XAbstractFeatureCall) expression, casted);
+				}
+			}
+		}
+		super.performMergeIntoParent();
+	}
+	
+	protected boolean canBeForwardResolved() {
+		if (expression instanceof XAbstractFeatureCall 
+				&& basicGetTypeParameters().isEmpty() 
+				&& basicGetTypeParameterHints().isEmpty() 
+				&& basicGetDeclardTypeParameters() == null
```

Comment:
Please double check this is correct. It is a bit surprising seeing a null check between all these "isEmpty" checks.

Reply:
In fact it is correct. `basicGetDeclaredTypeParameters` is `@Nullable` while the others are not.